src/test/pybind/test_cephfs.py

```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+        only empty directories.
+        '''
+        should_cancel = lambda: False
+
+        cephfs.mkdir('dir4', 0o755)
+        for i in range(1, 6):
+            cephfs.mkdir(f'/dir4/dir4{i}', 0o755)
+
+        cephfs.rmtree('dir4', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir4')
+
+    def test_rmtree_when_root_is_empty_dir(self, testdir):
+        '''
+        Test that rmtree() successfully deletes entire file hierarchy when it is
+        only an empty directory.
+        '''
+        should_cancel = lambda: False
+
+        cephfs.mkdir('dir5', 0o755)
+
+        cephfs.rmtree('dir5', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir5')
+
+    def get_file_count(self, dir_path):
+        '''
+        Return the number of files present in the given directory.
+        '''
+        i = 0
+        with cephfs.opendir(dir_path) as dir_handle:
+            de = cephfs.readdir(dir_handle)
+            while de:
+                i += 1
+                de = cephfs.readdir(dir_handle)
+        return i
+
+    def test_rmtree_no_perm_on_nonroot_dir(self, testdir):
+        '''
+        Test that rmtree() successfully deletes the entire file hierarchy except the
+        branch where permission for one of the (non-root) directories is not
+        granted.
+        '''
+        should_cancel = lambda: False
+
+        cephfs.mkdir('dir1', 0o755)
+
+        cephfs.mkdir('dir1/dir2', 0o755)
+        for i in range(1, 6):
+            fd = cephfs.open(f'/dir1/dir2/file{i}', 'w', 0o755)
+            cephfs.write(fd, b'abcd', 0)
+            cephfs.close(fd)
+
+        cephfs.mkdir('dir1/dir3', 0o755)
+        for i in range(1, 6):
+            fd = cephfs.open(f'/dir1/dir3/file{i}', 'w', 0o755)
+            cephfs.write(fd, b'abcd', 0)
+            cephfs.close(fd)
+
+        cephfs.mkdir('dir1/dir4', 0o755)
+        for i in range(1, 6):
+            fd = cephfs.open(f'/dir1/dir4/file{i}', 'w', 0o755)
+            cephfs.write(fd, b'abcd', 0)
+            cephfs.close(fd)
+
+        # actual test
+        cephfs.chmod('/dir1/dir3', 0o000)
+        cephfs.rmtree('dir1', should_cancel)
+        # ensure /dir1/dir3 wasn't deleted
+        cephfs.stat('dir1/dir3')
+
+        # cleanup
+        cephfs.chmod('/dir1/dir3', 0o755)
+        cephfs.rmtree('dir1', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir1')
+
+    def test_rmtree_no_perm_on_root(self, testdir):
+        '''
+        Test rmtree() exits when permission is not granted for the root of the file
+        hierarchy.
+        '''
+        should_cancel = lambda: False
+
+        cephfs.mkdir('dir1', 0o755)
+        for i in range(1, 6):
+            fd = cephfs.open(f'/dir1/file{i}', 'w', 0o755)
+            cephfs.write(fd, b'abcd', 0)
+            cephfs.close(fd)
+
+        cephfs.chmod('/dir1', 0o000)
+        cephfs.rmtree('dir1', should_cancel)
```

Comment:
This doesn't throw an exception?

Reply:
We have a separate test now to cover this case (since we can switch on and off whether exception is raised).